File: benchmarks/pandas/bench_multi_index_fromproduct.py
"""Benchmark: MultiIndex.from_product (pandas equivalent)."""
import json
import time
import pandas as pd

WARMUP = 3
ITERATIONS = 30

level1 = [f"a{i}" for i in range(50)]
level2 = list(range(100))

for _ in range(WARMUP):
    pd.MultiIndex.from_product([level1, level2])

t0 = time.perf_counter()
for _ in range(ITERATIONS):
    pd.MultiIndex.from_product([level1, level2])
total = (time.perf_counter() - t0) * 1000

print(json.dumps({"function": "multi_index_fromproduct", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))

File: benchmarks/tsb/bench_multi_index_fromproduct.ts
/**
 * Benchmark: MultiIndex.fromProduct — build from Cartesian product.
 */
import { MultiIndex } from "../../src/index.js";

const WARMUP = 3;
const ITERATIONS = 30;

const level1 = Array.from({ length: 50 }, (_, i) => `a${i}`);
const level2 = Array.from({ length: 100 }, (_, i) => i);

for (let i = 0; i < WARMUP; i++) {
  MultiIndex.fromProduct([level1, level2]);
}

const t0 = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  MultiIndex.fromProduct([level1, level2]);
}
const total = performance.now() - t0;

console.log(
  JSON.stringify({ function: "multi_index_fromproduct", mean_ms: total / ITERATIONS, iterations: ITERATIONS, total_ms: total }),
);
